src/hookintrprtr.c

```diff
@@ -0,0 +1,381 @@
+{
+    Int i;
+
+    CheckPrintOverflowWarnings();
+
+    if (HookActiveCount == HookCount) {
+        return 0;
+    }
+
+    HashLock(&activeHooks);
+    for (i = 0; i < HookCount; ++i) {
+        if (activeHooks[i] == hook) {
+            HashUnlock(&activeHooks);
+            return 0;
+        }
+    }
+
+    for (i = 0; i < sizeof(ExecStatFuncs) / sizeof(ExecStatFuncs[0]); i++) {
+        ExecStatFuncs[i] = ProfileStatPassthrough;
+        EvalExprFuncs[i] = ProfileEvalExprPassthrough;
+        EvalBoolFuncs[i] = ProfileEvalBoolPassthrough;
+    }
+
+    for (i = 0; i < HookCount; ++i) {
+        if (!activeHooks[i]) {
+            activeHooks[i] = hook;
+            HookActiveCount++;
+            HashUnlock(&activeHooks);
+            return 1;
+        }
+    }
+    HashUnlock(&activeHooks);
+    return 0;
+}
+
+Int DeactivateHooks(struct InterpreterHooks * hook)
+{
+    Int i;
+
+    HashLock(&activeHooks);
+    for (i = 0; i < HookCount; ++i) {
+        if (activeHooks[i] == hook) {
+            activeHooks[i] = 0;
+            HookActiveCount--;
+        }
+    }
+
+    if (HookActiveCount == 0) {
+        for (i = 0; i < sizeof(ExecStatFuncs) / sizeof(ExecStatFuncs[0]);
+             i++) {
+            ExecStatFuncs[i] = OriginalExecStatFuncsForHook[i];
+            EvalExprFuncs[i] = OriginalEvalExprFuncsForHook[i];
+            EvalBoolFuncs[i] = OriginalEvalBoolFuncsForHook[i];
+        }
+    }
+
+    HashUnlock(&activeHooks);
+    return 1;
+}
+
+/****************************************************************************
+**
+** These variables store if we have overflowed either 2^16 lines, or files.
+** In this case GAP will stop marking the line and file of statements.
+** We print a warning to tell users that this happens, either immediately
+** (if hooking is active) or when hooking is next activated if not.
+**/
+
+Int HaveReportedLineProfileOverflow;
+Int ShouldReportLineProfileOverflow;
+
+Int HaveReportedFileProfileOverflow;
+Int ShouldReportFileProfileOverflow;
+
+// This function only exists to allow testing of these overflow checks
+Obj FuncCLEAR_PROFILE_OVERFLOW_CHECKS(Obj self)
+{
+    HaveReportedLineProfileOverflow = 0;
+    ShouldReportLineProfileOverflow = 0;
+
+    HaveReportedFileProfileOverflow = 0;
+    ShouldReportFileProfileOverflow = 0;
+
+    return 0;
+}
+
+void CheckPrintOverflowWarnings()
+{
+    if (!HaveReportedLineProfileOverflow && ShouldReportLineProfileOverflow) {
+        HaveReportedLineProfileOverflow = 1;
+        Pr("#I Profiling only works on the first 65,535 lines of each file\n"
+           "#I (this warning will only appear once).\n",
+           0L, 0L);
+    }
+
+    if (!HaveReportedFileProfileOverflow && ShouldReportFileProfileOverflow) {
+        HaveReportedFileProfileOverflow = 1;
+        Pr("#I Profiling only works for the first 65,535 read files\n"
+           "#I (this warning will only appear once).\n",
+           0L, 0L);
+    }
+}
```

Comment:
Can we add a test for this? (e.g. by reading a string-stream with 65536 lines that we create on the fly)?

I wonder how the error would look. In particular, how can the user find out which files cause the problem? Right now, it would seem to me there is no way to do so at all?